probe-rs/src/architecture/arm/memory/adi_v5_memory_interface.rs

```diff
@@ -459,97 +554,224 @@ where
+
+        self.write_word_32(ap, Dcrsr::ADDRESS, dcrsr_val.into())
+            .unwrap();
+
+        self.wait_for_core_register_transfer(ap, Duration::from_millis(100))?;
+
+        let value = self.read_word_32(ap, Dcrdr::ADDRESS).unwrap();
+
+        Ok(value)
     }
 
-    fn read_word_8(&mut self, address: u32) -> Result<u8, Error> {
-        ADIMemoryInterface::read_word_8(self, address).map_err(Error::architecture_specific)
+    fn write_core_reg(
+        &mut self,
+        ap: MemoryAP,
+        addr: CoreRegisterAddress,
+        value: u32,
+    ) -> Result<(), Error> {
+        self.write_word_32(ap, Dcrdr::ADDRESS, value).unwrap();
+
+        // write the DCRSR value to select the register we want to write.
+        let mut dcrsr_val = Dcrsr(0);
+        dcrsr_val.set_regwnr(true); // Perform a write.
+        dcrsr_val.set_regsel(addr.into()); // The address of the register to write.
+
+        self.write_word_32(ap, Dcrsr::ADDRESS, dcrsr_val.into())
+            .unwrap();
+
+        self.wait_for_core_register_transfer(ap, Duration::from_millis(100))?;
+
+        Ok(())
     }
 
-    fn read_32(&mut self, address: u32, data: &mut [u32]) -> Result<(), Error> {
-        ADIMemoryInterface::read_32(self, address, data).map_err(Error::architecture_specific)
+    fn read_8(&mut self, ap: MemoryAP, address: u32, data: &mut [u8]) -> Result<(), Error> {
+        if data.len() == 1 {
+            data[0] = self.read_word_8(ap, address).unwrap();
+        } else {
+            self.read_8(ap, address, data).unwrap();
+        }
+
+        Ok(())
     }
 
-    fn read_8(&mut self, address: u32, data: &mut [u8]) -> Result<(), Error> {
-        ADIMemoryInterface::read_8(self, address, data).map_err(Error::architecture_specific)
+    fn read_32(&mut self, ap: MemoryAP, address: u32, data: &mut [u32]) -> Result<(), Error> {
+        if data.len() == 1 {
+            data[0] = self.read_word_32(ap, address).unwrap();
+        } else {
+            self.read_32(ap, address, data).unwrap();
+        }
+
+        Ok(())
+    }
+
+    fn write_8(&mut self, ap: MemoryAP, address: u32, data: &[u8]) -> Result<(), Error> {
+        if data.len() == 1 {
+            self.write_word_8(ap, address, data[0]).unwrap();
+        } else {
+            self.write_8(ap, address, data).unwrap();
+        }
+
+        Ok(())
+    }
+
+    fn write_32(&mut self, ap: MemoryAP, address: u32, data: &[u32]) -> Result<(), Error> {
+        if data.len() == 1 {
+            self.write_word_32(ap, address, data[0]).unwrap();
+        } else {
+            self.write_32(ap, address, data).unwrap();
+        }
+
+        Ok(())
+    }
+
+    fn flush(&mut self) -> Result<(), Error> {
+        self.interface.flush()?;
+
+        Ok(())
     }
+}
 
-    fn write_word_32(&mut self, address: u32, data: u32) -> Result<(), Error> {
-        ADIMemoryInterface::write_word_32(self, address, data).map_err(Error::architecture_specific)
+bitfield! {
+    #[derive(Copy, Clone)]
```

Comment:
Sadly this register is not the same for v6 and v7. So having it here is dangerous.

I am not sure what the best course of action would be for this.

Reply:
*v7:*

![image](https://user-images.githubusercontent.com/529795/97364169-f1c6b700-18a3-11eb-917e-0ac5434d2ab6.png)

*v6:*

![image](https://user-images.githubusercontent.com/529795/97364211-fee3a600-18a3-11eb-8237-5eaa8b28cfda.png)

Afaik v6 is completely compatible with v7. And as long as we don't use any v7 features, it's no issue.

Reply:
I think for now it's fine like this, we don't use the additional bit right now.